### PR TITLE
Replace coin by range chunking

### LINKED ISSUE #85: We can replace chamber's block to prime style
## Description

We can replace chamber's SMT and UTXO to prime style to handle `amount`.

https://ethresear.ch/t/plasma-prime-design-proposal/4222?u=syuhei176

- Chamber's SMT to SMT and RSA accumulator
- All coin has segment.

current chamber's style

```
{
owners: [0x00ab....],
value: [12345, 12346],
state: [OwnState]
}
```

after style

```
{
owners: [0x00ab....],
value: [{begin:12345, end:123456}],
state: [OwnState]
}
```

### PR DESCRIPTION
**Pull request by syuhei176**
## Description

Closes #85 

- [x] transfer

range chunking
https://ethresear.ch/t/plasma-prime-design-proposal/4222

**This is chamber style. not prime**

- We separate chunk with CHUNK_SIZE(ex: 1 ether)
- Merkle tree is ordered by slot.
- Tx is included in floor(start / CHUNK_SIZE) to floor(end / CHUNK_SIZE)

### CODE REVIEW

File: childchain/lib/block.js [hunk 1]
```diff
@@ -44,8 +46,12 @@ class Block {
     let leaves = Array.from(Array(Math.pow(2, SMT_DEPTH)), (item, index) => null);
     this.txs.forEach(tx=>{
       tx.outputs.forEach((o) => {
-        o.value.forEach((uid) => {
-          leaves[uid] = tx.hash();
+        o.value.forEach(({start, end}) => {
+          const slot = start.div(CHUNK_SIZE);
```
Comment (shogochiai) [inline childchain/lib/block.js]: `.integerValue(BigNumber.ROUND_FLOOR).toNumber();` ?

File: childchain/lib/block.js [hunk 2]
```diff
@@ -44,8 +46,12 @@ class Block {
...
+          const slotEnd = end.div(CHUNK_SIZE);
```
Comment (shogochiai) [inline childchain/lib/block.js]: `.integerValue(BigNumber.ROUND_FLOOR).toNumber();` ?

File: rootchain/test/RootChain.test.js [hunk 3]
```diff
@@ -208,9 +162,12 @@ contract('RootChain', function ([user, owner, recipient, user4, user5]) {
       assert.equal(
         r2[0], utils.bufferToHex(rootHash2)
       )
-      assert.equal(Merkle.verify(tx11.hash(), tx11.outputs[0].value[0], new Buffer(r1[0].substr(2), 'hex'), proof1), true);
-      assert.equal(Merkle.verify(tx21.hash(), tx21.outputs[0].value[0], rootHash2, proof2), true);
-      assert.equal(Merkle.verify(tx21.hash(), tx21.outputs[0].value[0], new Buffer(r2[0].substr(2), 'hex'), proof2), true);
+      const slot = tx11.outputs[0].value[0].start.div(CHUNK_SIZE);
```
Comment (shogochiai) [inline rootchain/test/RootChain.test.js]: `.integerValue(BigNumber.ROUND_FLOOR).toNumber();` ?

File: rootchain/test/RootChain.test.js [hunk 4]
```diff
@@ -252,7 +209,7 @@ contract('RootChain', function ([user, owner, recipient, user4, user5]) {
     });
 
     it('should startExit', async function () {
-      const slot = tx22.outputs[0].value[0];
+      const slot = tx22.outputs[0].value[0].start.div(CHUNK_SIZE).toNumber();
```
Comment (shogochiai) [inline rootchain/test/RootChain.test.js]: `.integerValue(BigNumber.ROUND_FLOOR).toNumber();` ?

File: childchain/lib/block.js [hunk 5]
```diff
@@ -60,17 +66,23 @@ class Block {
 
   createTXOProof(txo) {
     const tree = this.createTree();
-    return txo.value.map((uid) => tree.proof(uid)).reduce((acc, p) => {
-      return Buffer.concat(acc.concat([p]))
+    return txo.value.map(({start, end}) => {
+      const slot = start.div(CHUNK_SIZE).integerValue(BigNumber.ROUND_FLOOR).toNumber();
+      return tree.proof(slot);
```
Comment (shogochiai) [inline childchain/lib/block.js]:
```suggestion
      const slotEnd = end.div(CHUNK_SIZE).integerValue(BigNumber.ROUND_FLOOR).toNumber();
      return [tree.proof(slot), tree.proof(slotEnd)];
```

Proof duplication must be optimized

File: rootchain/test/RootChain.test.js [hunk 6]
```diff
@@ -28,18 +29,18 @@ contract('RootChain', function ([user, owner, recipient, user4, user5]) {
...
   const testAddress5 = utils.privateToAddress(privKey5);
   const zeroAddress = new Buffer("0000000000000000000000000000000000000000", 'hex');
-  const coin1Id = 1;
-  const coin2Id = 2;
+  const CHUNK_SIZE = BigNumber('1000000000000000000');
+  const coin1Id = {start: 0, end: CHUNK_SIZE};
+  const coin2Id = {start: CHUNK_SIZE, end: CHUNK_SIZE.times(2)};
   const gasLimit = 200000;
-  const startExitgasLimit = 600000;
+  const startExitgasLimit = 800000;
```
Comment (shogochiai) [inline rootchain/test/RootChain.test.js]: startExitGasLimit

File: rootchain/test/RootChain.test.js [hunk 7]
```diff
@@ -28,18 +29,18 @@ contract('RootChain', function ([user, owner, recipient, user4, user5]) {
   const testAddress4 = utils.privateToAddress(privKey4);
   const testAddress5 = utils.privateToAddress(privKey5);
   const zeroAddress = new Buffer("0000000000000000000000000000000000000000", 'hex');
-  const coin1Id = 1;
-  const coin2Id = 2;
+  const CHUNK_SIZE = BigNumber('1000000000000000000');
+  const coin1Id = {start: 0, end: CHUNK_SIZE};
```
[the remark this answers is not in the document]
Comment (syuhei176) [reply rootchain/test/RootChain.test.js]: I think this is a segment.

File: rootchain/test/TxVerification.test.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+const RLP = require('rlp');
+
+contract('TxVerificationTest', function ([user, owner, recipient, user4, user5]) {
+
+  const privKey1 = new Buffer('c87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3', 'hex')
+  const privKey2 = new Buffer('ae6ae8e5ccbfb04590405997ee2d52d2b330726137b875053c36d94e974d162f', 'hex')
+  const testAddress1 = utils.privateToAddress(privKey1);
+  const testAddress2 = utils.privateToAddress(privKey2);
+  const CHUNK_SIZE = BigNumber('1000000000000000000');
+  const coin1Id = {start: 0, end: CHUNK_SIZE.minus(1)};
```
[the remark this answers is not in the document]
Comment (syuhei176) [reply rootchain/test/TxVerification.test.js]: It's just mistake, fix now.

**Pull request merged**